Fixes issues w/ concurrent schema changes

tinwelint opened the pull request:
Individual commits tells the story, but generally this is about various issues with concurrent schema changes, such that the loaded representation of the schema may end up in state of permanently missing some updates, and also temporary reading temporarily broken schema.

Some of these issues has existed since the introduction of schema, but the more severe ones was introduced with the global schema lock split.

The `ConcurrentCreateDropIndexIT` test could previously run into all these issues in one or two runs, consistently, so could reproduce very reliably. For that reason I opted for adding only that test for this.

tinwelint committed:
Race has ability to add contestans that run max number of times

tinwelint committed:
Synchronizes createIndexes and dropIndex in IndexingService

because now these, and their updating the IndexMapReference by snapshot,
aren't protected by the global schema lock anymore. Now schema locking is
done per label.

tinwelint committed:
Fixes broken concurrency in SchemaCache

Previously there were only some data structures (maps/sets) that used implementations
which could handle concurrent access. Seemingly that was added ad-hoc.
Originally the state in this class was meant to not care about concurrency
and instead rely on external synchronization. However this was never fully
the case since there could be unguarded reads happening concurrently with updates
to these non-concurrency-aware data structures, which means that some existing
data could be observed as missing when other unrelated data items were removed
or added. This was indeed proven by ConcurrentCreateDropIndexIT in an upcoming
commit.

tinwelint committed:
More robust loading of all schema rules

Previously a load of all schema rule, e.g. when dropping an index rule
could temporarily observe deleted records and fail mid-way, this due to
another concurrent drop. This race opportunity was likely made available
when breaking the global schema lock into per-label locks.
The loading is now aware of this and ignores unused records it may see
along the way. It still fails on malformed records.

tinwelint committed:
Tests for stressing concurrent create and drop indexes

This exercises concurrency logic that was recently made more racy by
breaking the global schema lock into per-label locks.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -437,7 +437,7 @@ private void apply( Iterable<IndexEntryUpdate<LabelSchemaDescriptor>> updates, I
      * will shut down.
      */
-    public void createIndexes( IndexRule... rules ) throws IOException
+    public synchronized void createIndexes( IndexRule... rules ) throws IOException
```
Only the `createIndexes` and the `dropIndex` synchronise on the `IndexingService` monitor lock as far as I can see. This means the state in here is inconsistently synchronised. If we are only worried about concurrent updates to the `indexMapRef`, then we should synchronise on that, and do so consistently. This synchronisation could, for instance, move into the `IndexMapReference` class and be exposed as an atomic `update` method that takes a lambda, or something like that.

tinwelint replied:
Maybe. Which are the other usages that should have been synchronized?

chrisvest replied:
It's not that `init` and `start` _need_ to synchronise. Rather, I'd like the synchronisation to be consistent, e.g. by pushing the responsibility down to the `IndexMapReference`, such that there is no way to get it wrong.

tinwelint replied:
Right OK, that's of course never a bad idea. I can have a look at that suggestion about a lambda, I like it.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/SchemaCache.java:
```diff
@@ -55,16 +52,16 @@
 
     private final Map<Class<?>,Object> dependantState = new ConcurrentHashMap<>();
-    private final PrimitiveIntObjectMap<List<IndexDescriptor>> indexByProperty = Primitive.intObjectMap();
+    private final Map<Integer,List<IndexDescriptor>> indexByProperty = new ConcurrentHashMap<>();
```
I think we should have a card for looking into an implementation of this based on snapshotting and atomic multi-value updates.

tinwelint replied:
I think so too.

chrisvest commented on community/kernel/src/test/java/org/neo4j/graphdb/schema/ConcurrentCreateDropIndexIT.java:
```diff
@@ -0,0 +1,110 @@
+    public final DatabaseRule db = new ImpermanentDatabaseRule();
+
+    @Test
+    public void shouldTest() throws Throwable
```
How about `concurrentCreatingAndDroppingOfIndexesShouldNotInterferre`?

tinwelint replied:
Forgot to rename, it started out as a quick-just-get-this-reproduced thing :)

chrisvest commented on community/kernel/src/test/java/org/neo4j/graphdb/schema/ConcurrentCreateDropIndexIT.java:
```diff
@@ -0,0 +1,110 @@
+            assertEquals( 0, asList( db.schema().getIndexes() ).size() );
+            tx.success();
+        }
+    }
```
I think it would be good to also have a middle phase where threads/2 indexes are dropped concurrently with another threads/2 indexes being created, asserting that the index count ends up unchanged.

tinwelint replied:
Good idea, I can add that

chrisvest reviewed (changes requested):
@tinwelint First round completed. You have a couple of failing tests, and I have a few comments.

tinwelint committed:
Improves on testing with more cases

tinwelint committed:
Bullet proof IndexMapReference modification

where it's only possible to modify its contents (with the existing snapshot
technique) by means of one synchronized method accepting a lambda containing
the logic for making the change.

tinwelint commented:
@chrisvest something like this?

chrisvest reviewed (approved):
Very nice 👍
